Inconsistent quotes

wes-turner opened the pull request:
Ports g-inconsistent-quotes to pylint.

Quoting PEP-8:
> In Python, single-quoted strings and double-quoted strings are the same. This PEP does not make a recommendation for this. Pick a rule and stick to it. When a string contains single or double quote characters, however, use the other one to avoid backslashes in the string. It improves readability.
> 
> For triple-quoted strings, always use double quote characters to be consistent with the docstring convention in PEP 257.

Features:
* Accounts for strings where the delimiter is swapped so an internal quote doesn't need to be escaped
* Only errors on those lines that represent the module's minority delimiter.
* Ignores longstrings (they could be docstrings, and checking those delimiters is someone else's responsibility)

wes-turner committed:
Simple (failing) test for inconsistent-quotes

Single example of inconsistent-quotes in functional_tests.

wes-turner committed:
Checkpoint for work in progress.

Currently throws more warnings than it should.

wes-turner committed:
Patches small bugs

* Fixed most-common-quote determination.  Now only errors on those lines
that are in the minority.
* Matches properly on multiline longstrings.

wes-turner committed:
inconsistent-quotes is now optional

Default is to not check.

wes-turner committed:
Updated tests for inconsistent-quotes.

* Added a docstring string
* Added a second file where single quotes are the majority delimiter.

wes-turner commented on pylint/checkers/strings.py:
```diff
@@ -672,6 +727,23 @@ def visit_assign(self, node):
             self.check_for_concatenated_strings([node.value], "assignment")
 
+    def check_for_consistent_string_delimiters(self, tokens):
+        # Check for consistent string delimiters
```
Give this a real docstring.

wes-turner committed:
Docstring for check_for_consistent_string_delimiters

wes-turner commented on pylint/checkers/strings.py:
```diff
@@ -672,6 +727,33 @@ def visit_assign(self, node):
         if isinstance(node.value, astroid.Const) and isinstance(node.value.value, str):
             self.check_for_concatenated_strings([node.value], "assignment")
 
+    def check_for_consistent_string_delimiters(
```
I don't like the dependency on work having been done in the caller to count delimiters.  The upside I see for that work being done there is that tokens are being iterated through there, already, for a variety of reasons.  What do you folks think about moving that logic into this method for its own pass through all the tokens?

wes-turner replied:
If that work is pulled into this method, we can dispense with `self._string_delimiters`, too.  That's a nice bonus.

wes-turner replied:
I'm just going to make that change.

wes-turner committed:
Moves quote-delimiter counting into the function that checks delimiters.

Counting used to take place in a general-purpose iteration through the
tokens.  Now it takes place only where it's needed -- in the function
that actually performs the check.

martindemello commented on pylint/checkers/strings.py:
```diff
@@ -787,3 +871,66 @@ def str_eval(token):
+    return (
+        quote_delimiter
+        and not is_long_string(token)
+        and not re.search(non_delimiter, str_eval(token))
```
you can just say `non_delimiter not in str_eval(token)` here

wes-turner replied:
Oh!  That's good to know.   Thanks!

martindemello commented on pylint/checkers/strings.py:
```diff
@@ -787,3 +871,66 @@ def str_eval(token):
+        A boolean representing whether or not this token matches a longstring
+        regex.
+    """
+    single_quoted_regex = "(%s)?'''.*'''\Z" % "|".join(_STRING_AND_BYTES_PREFIXES)
```
do you need to check the complete string (as opposed to just the prefix) here?

wes-turner replied:
Mmm.  You're totally right.  The only thing we get from checking the rest of the string is some extra validation that the token is well-formed.

But checking well-formedness isn't really this function's job, and per your point, the extra code obscures what the function's purpose is.  So I'm going to make the change and add a little description in the docstring about well-formedness non-guarantees.

wes-turner replied:
Ok, now I'm second-guessing this decision.  The function isn't `has_long_string_prefix`, it claims to be `is_long_string.`  I want a function that purports to make that sort of judgment check for well-formedness.

So maybe the next question is to ask whether the implemented functionality actually needs an `is_long_string` function, and maybe if it should call a `has_long_string_prefix` instead.  There... I dunno.  I think if we need to write code anyway, then writing a function like `is_long_string` is both *very* slightly more useful and a little easier to mentally chunk.

So I think my inclination is to leave this code as-is.  What are your thoughts?

martindemello replied:
my take is that a string that is not well-formed will not pass the python compiler, and therefore we can assume that a string starting with `'''` will necessarily end with `'''`. so keeping the function as `is_long_string` but only checking the prefix is fine here.

wes-turner replied:
Ok.  I think that's a good argument.  I'll go in and make the change.

martindemello commented on pylint/checkers/strings.py:
```diff
@@ -787,3 +871,66 @@ def str_eval(token):
+        returned instead.
+    """
+    match = re.match(
+        "(%s)?(\"|')" % "|".join(_STRING_AND_BYTES_PREFIXES), token, re.DOTALL
```
consider compiling all these regular expressions as globals, just after you define `_STRING_AND_BYTES_PREFIXES`

wes-turner replied:
I'm torn, here.

I expect our end decision to be purely about readability.  I haven't run any benchmarks, but per [the docs](https://docs.python.org/3/library/re.html#re.compile):

> the module-level matching functions are cached, so programs that use only a few regular expressions at a time needn’t worry about compiling regular expressions.

I generally prefer pulling strings and other magic numbers out from functions to module-level.  I like how it encourages me to refer to them by name rather by literal and give them more meaning in so doing.  I like how it separates logic and instance details, helping a reader focus on both independently.

But here, I find myself not wanting to do that.  The instance details seem more tightly coupled with the logic.  When I read `match.group(2)`, I need to look at the regex in question to understand what it refers to.  That's easier when the regex is right there than if it were at the top of the file.  The long_string regexes aren't quite so bad, in this respect -- if, even after reading this, you think it reads better to have those up top, I'd happily move them.

I couldn't find any guidance in the style guide, but did find this from Nathaniel in g/python-style:

> It is never wrong to move a code element that can be at module scope from function scope to module scope. It is only sometimes wrong to refrain from doing so. Some authors like the locality of having constants close to their one use-site; other authors like the way that migrating to module scope everything that can be so migrated shrinks and simplifies their function implementations.

What do you think about, yourself, when making this sort of judgment call?

martindemello replied:
i didn't know the runtime cached calls to `re.compile`! the existing code should be fine, then. you could do the join as a global variable if you like, so your code would be `re.match("(%s)?..." % _STRING_AND_BYTES_PREFIXES)` without the need to repeat the string join everywhere, more as a matter of the code looking uncluttered than performance. (i also think a shorter name like `_STRING_PREFIXES` or even just `_PREFIXES` given that the file is `strings.py` would be fine, though that's just a matter of personal taste)

wes-turner replied:
I like the implicit context with the name `_PREFIXES` a lot.  I've made that change.

You're right that it would unclutter the code, but I've held off on creating a global representing the string in `"|".join(_PREFIXES)` because that string seems like a combination of technical implementation and instance.  I can't think of a clear and succinct meaning it has as a string, separated from its use in the regex.  In this case, I prefer the clutter because of that clearer meaning.

wes-turner committed:
Idiomatic reformulation of re.search.

Changed a re.search to be more Pythonic.

wes-turner committed:
Added documentation for new 'inconsistent-quotes' check.

wes-turner committed:
Small code-review changes.

wes-turner committed:
Explicitly casts return types

New return types fit the type annotations exactly.

wes-turner committed:
get_quote_delimiter returns str rather than Optional[str]

If no quote delimiter was found (because, probably, the input was not a
well-formed string), raise an error rather than return None.

wes-turner committed:
Marks internal functions as private.
